src/dune/build_system.ml

```diff
@@ -450,7 +450,30 @@ let compute_targets_digest_or_raise_error ~loc targets =
   let good, bad =
     List.partition_map targets ~f:(fun target ->
         let fn = Path.build target in
-        match Cached_digest.refresh fn with
+        let remove_write_permissions =
+          (* Remove write permissions on targets. A first theoretical reason is
+             that the build process should be a computational graph and targets
+             should not change state once built. A very practical reason is that
+             enabling the cache will remove write permission because of hardlink
+             sharing anyway, so always removing them enables to catch mistakes
+             earlier. *)
+          (* FIXME: searching the dune version for each single target seems way
+             suboptimal. This information could probably be stored in rules
+             directly. *)
+          let _, src_dir = Path.Build.extract_build_context_dir_exn target in
+          let dir = File_tree.nearest_dir src_dir in
+          let version =
+            File_tree.Dir.project dir |> Dune_project.dune_version
+          in
+          version >= (2, 4)
+        in
+        let refresh =
+          if remove_write_permissions then
+            Cached_digest.refresh_and_chmod
+          else
+            Cached_digest.refresh
+        in
```

Comment:
The above part can go outside of the `List.partition_map` since targets are always in the same directory. We plan to lift this restriction, but the targets will at least always be part of the same project.